test/TEST-10-ISSUE-2467/test.sh

```diff
@@ -29,13 +28,15 @@ After=multi-user.target
 
 [Service]
 Type=oneshot
-ExecStart=/bin/sh -e -x -c 'rm -f /tmp/nonexistent; systemctl start test.socket; echo a | nc -U /run/test.ctl; >/testok'
+StandardOutput=tty
+StandardError=tty
+ExecStart=/bin/sh -e -x -c 'rm -f /tmp/nonexistent; systemctl start test.socket; echo > /run/test.ctl; >/testok'
```

Comment:
The test doesn't seem to test what it is supposed to test. This could be checked by adding `TriggerLimitIntervalSec=0` to `test.socket` to imitate the bug fixed in 8b26cdbd2a949b02c0f4d.

I think it would be better to use `socat` instead of `nc` or simply ignore the exit code of `nc` by using something like `|| true`.

Reply:
hmm, can you elaborate on this? what does the pipe-based solution do differently from the unix-based solution?

any chance you can rewrite this with socat? it appears socat is the better choice for this, since it doesn't vary that widely in its implementation

Reply:
`echo` always writes data successfully whatever happens. `nc`, on the other hand, is blocked until it receives a reply, which helps to detect how long it takes `systemd` to activate the service. If it takes more that 10 seconds, the service fails because it has `TimeoutStartSec=10s`.

Regarding `socat`, @Siosm has already written [the code where `socat` is used](https://github.com/Siosm/systemd/commit/5aabe70e456173de06642a0fd4e0ae85ea16d40f), but hasn't created a PR yet.

Reply:
Sorry for resurrecting such an old thread, but, in theory, shouldn't be enough to use `/bin/echo` instead of just `echo`? From what I remember, in some implementations[0] the `echo` builtin doesn't check if the `write()` syscall succeeded, however `/bin/echo` from coreutils does.

[0] 
RHEL 7.6, bash-4.2.46-31.el7.x86_64
```
$ man bash
...
        echo [-neE] [arg ...]
              Output  the  args,  separated by spaces, followed by a newline.  The return status is always 0.
```

However:
Fedora 28, bash-4.4.23-1.fc28.x86_64
```
$ man bash
...
       echo [-neE] [arg ...]
              Output the args, separated by spaces, followed by a newline.  The return status is 0 unless a  write  error  occurs.
```